pkg/pligos/compiler/compile.go

```diff
@@ -1,95 +1,52 @@
+	schema    map[string]interface{} // describes the schema for the output
+	types     map[string]interface{} // describes all available types
+	instances map[string]interface{} // describes all instances
 }
 
-type ValuesEncoderMap map[string]*ValuesEncoder
-
-func NewValuesEncoderMap(contexts []Context, types map[string]interface{}, values map[string]interface{}) (ValuesEncoderMap, error) {
-	res := make(map[string]*ValuesEncoder)
-	for _, context := range contexts {
-		schema, err := CreateSchema(filepath.Join(context.Flavor, "schema.yaml"), types)
-		if err != nil {
-			return nil, err
-		}
-
-		res[context.Name] = NewValuesEncoder(values, schema)
-	}
-
-	return ValuesEncoderMap(res), nil
-}
-
-func (v ValuesEncoderMap) Get(context string) *ValuesEncoder {
-	return v[context]
-}
-
-func NewValuesEncoder(config map[string]interface{}, schema map[string]interface{}) *ValuesEncoder {
-	return &ValuesEncoder{
-		config: config,
-		schema: schema,
+func New(config, schema, types, instances map[string]interface{}) *Compiler {
+	return &Compiler{
+		config:    config,
+		schema:    schema,
+		types:     types,
+		instances: instances,
 	}
 }
 
-func (ve *ValuesEncoder) repeatedLength(in interface{}) int {
+func (ve *Compiler) repeatedLength(in interface{}) int {
 	if s, ok := in.([]map[string]interface{}); ok {
 		return len(s)
 	}
 
 	return len(in.([]interface{}))
 }
 
-func (ve *ValuesEncoder) handleMappedRoot(context, mappedType string) (map[string]interface{}, error) {
-	res := make(map[string]interface{})
-	typ, err := ve.resolveType(mappedType)
-	if err != nil {
-		return nil, err
-	}
-
-	for _, e := range ve.config["contexts"].(map[string]interface{})[context].([]interface{}) {
-		i, err := ve.resolveTypeInstance(mappedType, e.(string))
-		if err != nil {
-			return nil, err
-		}
-		n, err := ve.Encode(typ, i)
-		if err != nil {
-			return nil, err
-		}
-		res[e.(string)] = n
-	}
-
-	return res, nil
-}
-
-func (ve *ValuesEncoder) EncodeContext(context string) (map[string]interface{}, error) {
-	if mappedType, ok := isOfType(ve.schema["context"], mapped); ok {
-		return ve.handleMappedRoot(context, mappedType)
-	}
-
-	return ve.Encode(
-		ve.schema["context"].(map[string]interface{}),
-		ve.config["contexts"].(map[string]interface{})[context].(map[string]interface{}),
+func (ve *Compiler) Compile() (map[string]interface{}, error) {
+	return ve.compile(
+		ve.schema,
+		ve.config,
 	)
 }
 
-func (ve *ValuesEncoder) embed(res map[string]interface{}, config map[string]interface{}) map[string]interface{} {
+func (ve *Compiler) embed(res map[string]interface{}, config map[string]interface{}) map[string]interface{} {
 	for k, v := range config {
 		res[k] = v
 	}
 
 	return res
 }
 
-func (ve *ValuesEncoder) Encode(schema map[string]interface{}, config map[string]interface{}) (map[string]interface{}, error) {
+func (ve *Compiler) compile(schema map[string]interface{}, config map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
I can't rename `typ` to `type` since it's a keyword in Go. `typ` is actually used in the standard library as well (look for instance at the `reflect` package). I would rather not change `typXyz` to `typeXyz` to stay consistent with `typ`